File: C/Lista5/Lista5.scala
/** Zadanie 2 */
val lfib = {
    def fib(number: Int, acc: Int): Stream[Int] = Stream.cons(number, fib(acc, number + acc))
  fib(0, 1)
}




/** Zadanie 3 */
sealed trait lBT[+A]
case object LEmpty extends lBT[Nothing]
case class LNode[+A](elem: A, left: () => lBT[A], right: () => lBT[A]) extends lBT[A]

// a)
def lTree(n: Int): lBT[Int] = LNode(n, () => lTree(2 * n), () => lTree(2 * n + 1))

// b)
def breadth[A](ltree: lBT[A]) = {
    def breadthHelp(el: List[lBT[A]]): Stream[A] = {
      el match {
        case Nil         => Stream.Empty
        case LEmpty :: t => breadthHelp(t)
        case LNode(v, l, r) :: t =>
          Stream.cons(v, breadthHelp(t ::: (l() :: r() :: Nil)))
      }
    }
  breadthHelp(List(ltree))
}

File: C/Lista5/Lista5.ml
type 'a llist = LNil | LCons of 'a * (unit -> 'a llist)

(* Zadanie 1 *)
let rec replicateElements (k, list) =
	let rec replicate = function
		| (_, LNil) -> LNil
		| (n, LCons(h, xf)) -> 
			if n = 0 then replicate(k, xf())
			else LCons(h, fun () -> replicate (n - 1, xf()))
	in replicate(k, list)
	
(* Zadanie 2 *)
let lfib =
	let rec fib (number, acc) =
		LCons(number, fun () -> fib(acc, number + acc))
	in fib(0, 1)

(* Zadanie 3 *)
type 'a lBT = LEmpty | LNode of 'a * (unit ->'a lBT) * (unit -> 'a lBT)

(* a) *)
let rec lTree n =
	LNode(n, (fun () -> lTree(2 * n)), (fun () -> lTree(2 * n + 1)))

(* b) *)
let breadth ltree =
	let rec breadthHelp = function
		| [] -> LNil
		| LEmpty :: t -> breadthHelp t
		| LNode(v, l, r) :: t ->
				LCons(v, fun () -> breadthHelp (t @ (l() :: r() :: [])))
	in breadthHelp [ltree]